server/ai_live_video.go

```diff
@@ -318,6 +324,39 @@ func copySegment(segment *http.Response, w io.Writer) (int64, error) {
 	return io.Copy(w, segment.Body)
 }
 
+func copySegmentWithTimeout(segment *http.Response, w io.Writer, timeout time.Duration) (int64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	// Create a buffer to store the data
+	// We don't copy directly into the writer to avoid partly written segments
+	buffer := &bytes.Buffer{}
```

Comment:
Would it be enough to directly write to `w` as before and have the timeout on that operation? It means we're not specifically timing out the reading part but does that matter, we could simplify things a bit and just have the timeout on both read and write?

Reply:
Yeah I'm not a fan of buffering this write either. We have a lot of complexity to guarantee a full streaming workflow with trickle and whatnot, so this would be beating the purpose of all that and buffering the whole segment instead, increasing latency. 

One alternative to have the timeout but still avoid partial writes is to check the timeout only on the _first read_ from the `segment.Body`, and only after that using `io.Copy` to copy the rest of the segment. If an error occurs on the first read, we can just err with the timeout. If it errs after the first read, we'd need to `w.CloseWithError` (if available, it exists on `io.Pipe` at least) to propagate a recover on the read side somehow.

Reply:
The reason why I added the buffer was that if we "re-use" the same pipe in the ffmpeg for O Swapping, then partial segment write will break the pipe. So, I wanted to avoid writing partial segments. Wdyt?

Also @j0sh , wdyt?

Reply:
One other option is to remove the buffering for now and maybe we can try to re-think it while actually introducing O Swapping. Maybe ffmpeg will be able to handle this partial corrupted segments.

Reply:
nvm, I think after second thought, I'll remove this buffer. You're right that if the segment is e.g. 1s long, then it would add additional 1s latency.

Reply:
Ok, removed buffer, left the timeout. PTAL @mjh1 @victorges